### ISSUE TITLE
Regression related to Improve lazy ghost performance by avoiding self-referencing closure

### ISSUE THREAD

**Issue by raziel057** [doctrine/doctrine2#11455]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.19.3

#### Summary

From the following commit of doctrine/orm 2.19.3 I got a OutOfMemoryError when trying to clone an entity that contains a blob value: https://github.com/doctrine/orm/commit/9d1a4973ae2d8ddedd6b66e958e32ebb58cdf3d6

If I revert this change I have no issue and the Memory is low: less than 4Mo but when I apply the change, even if I allocate 2G of memory I got the OutOfMemoryError .

#### Current behavior

Infinite memory rise which leads to an OutOfMemoryError.

#### How to reproduce

```php
namespace App\Entity;

...

#[ORM\Entity(repositoryClass: ResourceFileRepository::class)]
class ResourceFile
{

    #[Groups(['read'])]
    #[ORM\Id]
    #[ORM\Column(name: 'ID', type: 'integer', nullable: false)]
    #[ORM\GeneratedValue(strategy: 'IDENTITY')]
    private ?int $id = null;

    ...

    /**
     * @var string|resource
     */
    #[ORM\Column(name: 'VALUE', type: 'blob', nullable: false)]
    private mixed $content = null;

    public function getCopy() : self
    {
        return clone $this;
    }
    
    public function __clone()
    {
        if ($this->id) {
            $this->id = null;
        }
    }
}
```

As soon as I call ``getCopy();`` I got an OutOfMemoryError:

```php
$copy = $formDefinition->getLogo()->getCopy();
```

![image](https://github.com/doctrine/orm/assets/652505/1b29fcf4-9075-4cb0-bcc5-5f651b921eb1)

Please note that I have no issue if I call ``$formDefinition->getLogo()->getContent()`` to load the resource before calling ``$formDefinition->getLogo()->getCopy()`` :

```php

dump("Initial Memory Used: ".round(memory_get_usage() / (1024*1024), 1)." MB");

$formDefinition = $this->em->getRepository(FormDefinition::class)->find($id);

dump($formDefinition->getLogo()->getContent());
dump("Memory Used before getCopy: ".round(memory_get_usage() / (1024*1024), 1)." MB");

$copy = $formDefinition->getLogo()->getCopy();

dump("Memory Used after getCopy: ".round(memory_get_usage() / (1024*1024), 1)." MB");
```

![image](https://github.com/doctrine/orm/assets/652505/4dbd41c0-564d-4462-b012-254b96ac85f9)

#### Expected behavior

Just clone the entity.

**Comment by raziel057** [comment]
@nicolas-grekas Can you please consider this issue as you created the commit https://github.com/doctrine/orm/commit/9d1a4973ae2d8ddedd6b66e958e32ebb58cdf3d6

If you need more information or to test something I'm available.

**Comment by nicolas-grekas** [comment]
A small repo with instructions to reproduce would help a lot :pray:

**Comment by raziel057** [comment]
I tried to create a small / simple reproducer, but could notice the issue is also related to the number of elements I have in the  ResourceFile table in database. If I reduce the number of elements I don't get the MemoryLimitError.

I also detected that I got the error if I load the parent entity. For example in the following sample The formDefinition entity contains a logo (ResourceFile id 24163). If I comment the first line I don't have any Memory issue, but if I uncomment it the memory issue happen.

```php
$formDefinition = $entityManager->getRepository(FormDefinition::class)->find(3829);
$resourceFile = $entityManager->getRepository(ResourceFile::class)->find(24163);

$copy = clone $resourceFile;
```

Here is the debug:

![image](https://github.com/doctrine/orm/assets/652505/8308e08c-1423-4f19-8f57-b95aeead7361)

And at some point a query is done without restricting on ID:

![image](https://github.com/doctrine/orm/assets/652505/3e570894-c22b-4bd1-bf20-945da1c8a9b4)

If in ``clone`` method of ``ResouyrceFile`` I remove the following code, I don't have any issue and no query are triggered:
```php
    if ($this->creationDate) {
        $this->creationDate = null;
    }
```

![image](https://github.com/doctrine/orm/assets/652505/9934a66a-895e-4083-a1f4-184c71871bc9)

![image](https://github.com/doctrine/orm/assets/652505/b5b295ea-af8f-4c61-8d31-2c52ce8cb5af)

The code is not very easy to read so I can't provide more information for now.

**Comment by raziel057** [comment]
@nicolas-grekas Here is a reproducer:

```
git clone https://github.com/raziel057/demo.git -b blob-issue
cd demo/
composer install
vendor/bin/simple-phpunit
```

Result

![image](https://github.com/doctrine/orm/assets/652505/00fc2d37-e126-4654-ba85-d7b35c3101cf)

Downgrade to doctrine/orm:2.19.2 and rerun phpunit:

```
composer require doctrine/orm:2.19.2
vendor/bin/simple-phpunit
```

![image](https://github.com/doctrine/orm/assets/652505/1e5ceaab-5055-4abe-8942-c84262f365eb)

As you can see in the first result the query doesn't contain the where restriction. That's why that was leading to an OutOfMemoryError in my case as I have a lot of entities in my table.

Can you please have a look now?

**Comment by nicolas-grekas** [comment]
Thanks for the reproducer. I already spotted a bug but not the one you reported :) Stay tuned.

**Comment by raziel057** [comment]
Nice 👍 thanks

**Comment by nicolas-grekas** [comment]
This should be fixed by #11475
